### PR TITLE
added basic models

### PR DESCRIPTION
**Pull request by BrainIsDead**

### CODE REVIEW

File: src/datasources/models.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+    name = models.CharField(
+        help_text=_('Name'),
+        max_length=128,
+        unique=True
+    )
+    base_name = models.CharField(
+        help_text=_('Signal BaseName'),
+        max_length=128,
+        unique=True
+    )
```
[the remark this answers is not in the document]
Comment (krivard) [reply src/datasources/models.py:125]: From discussion over zoom:

* one base name can apply to many signals; each signal _currently_ has only one base name
* it's not always clear which signal should be the base name among a group of related signals (example: fb-survey signals used to use the unweighted signal as base; now use weighted signal as base)
* prototype sheet already has {base_name} {base_description} etc populated in the display name and description text; not sure how we want to handle this in the app (options: (1) use fully-populated text in the app; (2) use & permit templated text in the app... (2) is probably better but we should write down how we want it to work so it's clear how to implement it)

File: src/datasources/models.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    base_is_other = models.BooleanField(
+        help_text=_('Base is other'),
+        default=False
+    )
```
[the remark this answers is not in the document]
Comment (BrainIsDead) [reply src/datasources/models.py:129]: Done

File: src/datasources/models.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    compute_from_base = models.BooleanField(
+        help_text=_('Compute From Base'),
+        default=False
+    )
```
[the remark this answers is not in the document]
Comment (BrainIsDead) [reply src/datasources/models.py]: Done

File: src/datasources/models.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+        max_length=128,
+        choices=TimeLabelChoices.choices
+    )
+    format = models.CharField(
+        help_text=_('Format'),
+        max_length=128,
+        choices=FormatChoices.choices
+    )
+    category = models.CharField(
+        help_text=_('Format'),
```
[the remark this answers is not in the document]
Comment (BrainIsDead) [reply src/datasources/models.py]: Done

File: src/datasources/models.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+    )
+    high_values_are = models.CharField(
+        help_text=_('High values are'),
+        max_length=128,
+        choices=HighValuesAreChoices.choices
+    )
+    link = models.URLField(
+        help_text=_('Link'),
+        max_length=256
+    )
```
[the remark this answers is not in the document]
Comment (BrainIsDead) [reply src/datasources/models.py]: Done, but will be changed to models

File: src/datasources/models.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+        help_text=_('Reference Signal'),
+        on_delete=models.PROTECT
+    )
+    pathogen = models.ForeignKey(
+        'Pathogen',
+        related_name='pathogen',
+        help_text=_('Pathogen/Disease Area'),
+        on_delete=models.SET_NULL,
+        null=True
+    )
```
[the remark this answers is not in the document]
Comment (BrainIsDead) [reply src/datasources/models.py]: Done

**Comment by BrainIsDead** [comment]
> Comments inline -- it feels like some of this code was maybe premature? I'm used to discussing data models in diagrams first so everyone's on the same page with what the relationships are and where they belong. Django is beautifully concise, but not as concise as a drawing 😄

models

File: src/base/models.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+class Link(models.Model):
+    """
+    A model representing a Link.
+    """
+    link_type = models.CharField(
+        help_text=_('Link type'),
+        choices=LinkTypeChoices.choices,
+        max_length=128,
+        unique=True
+    )
```
Comment (krivard) [inline src/base/models.py:24]: We need to support the following anchor text on links[1]:

* DUA
* API Documentation
* Interpreting mask use in context
* Question text
* Survey details
* Survey documentation
* Technical description
* Wave 10 revision updates
* Wave 11 revision updates

Will `LinkType` do this, or does Django support a way to include an extra column on the many-to-many table(s) between Link and the things-that-have-links, or do we need something else?

[1] I pulled a unique list of anchor text from the Source and Signals sheets into [a scratch page on the MVP sheet](https://docs.google.com/spreadsheets/d/168ePK0XAzlGKvtYItpQlX6Gh0CHlLKNqd185GYWEA24/edit#gid=792762620)

**Review by krivard — CHANGES_REQUESTED** [review]
Some cleanup, a question for Carlyn, and a question about LinkTypes

File: src/signals/models.py [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+        on_delete=models.SET_NULL,
+        null=True
+    )
+    signal_type = models.ForeignKey(
+        'signals.SignalType',
+        related_name='signals',
+        help_text=_('Signal Type'),
+        on_delete=models.SET_NULL,
+        null=True
+    )
```
[the remark this answers is not in the document]
Comment (carlynvandyke) [reply src/signals/models.py]: Just discussed this with Mike- we are going to make it many-to-many.

File: src/base/models.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+class LinkTypeChoices(models.TextChoices):
+    """
+    A class representing choices for link types.
+    """
+    API_DOCUMENTATION = 'api_documentation', _('API Documentation')
+    DUA = 'dua', _('DUA')
+    INTERPRETING_MASK = 'interpreting_mask', _('Interpreting mask use in context')
+    QUESTION_TEXT = 'question_text', _('Question text')
+    SURVEY_DETAILS = 'survey_details', _('SURVEY_DETAILS')
```
Comment (krivard) [inline src/base/models.py:13]:
```suggestion
    SURVEY_DETAILS = 'survey_details', _('Survey details')
```

File: src/datasources/models.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    links = models.ManyToManyField(
+        'base.Link',
+        help_text=_('Source Subdivision links'),
+        related_name="source_subdivisions"
+    )
+    data_source = models.ForeignKey(
+        'datasources.DataSource',
+        related_name='source_subdivisions',
+        help_text=_('Source Subdivision'),
+        on_delete=models.SET_NULL,
```
Comment (krivard) [inline src/datasources/models.py:39]: use PROTECT here: the system shouldn't let you delete a DataSource until its subdivisions have either been deleted or reassigned

I would also accept CASCADE but I don't feel strongly either way; use your favorite

File: src/signals/models.py [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    """
+    name = models.CharField(
+        help_text=_('Name'),
+        max_length=128,
+        unique=True
+    )
+    base = models.ForeignKey(
+        'signals.Signal',
+        help_text=_('Signal base'),
+        on_delete=models.SET_NULL,
```
Comment (krivard) [inline src/signals/models.py:136]: use PROTECT here: the system should not let you delete a signal until its related signals have been reassigned

File: src/signals/models.py [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+        null=True
+    )
+    pathogen = models.ForeignKey(
+        'signals.Pathogen',
+        related_name='signals',
+        help_text=_('Pathogen/Disease Area'),
+        on_delete=models.SET_NULL,
```
Comment (krivard) [inline src/signals/models.py:143]: confirming SET_NULL: if a pathogen is deleted, the signals for that pathogen should be reassigned to the null pathogen

File: src/signals/models.py [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+    time_label = models.CharField(
+        help_text=_('Time label'),
+        max_length=128,
+        choices=TimeLabelChoices.choices
+    )
+    category = models.ForeignKey(
+        'signals.SignalCategory',
+        related_name='signals',
+        help_text=_('Signal Category'),
+        on_delete=models.SET_NULL,
```
Comment (krivard) [inline src/signals/models.py:186]: confirming SET_NULL: if a category is deleted, the signals in that category should be reassigned to the null category

File: src/signals/models.py [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+    high_values_are = models.CharField(
+        help_text=_('High values are'),
+        max_length=128,
+        choices=HighValuesAreChoices.choices
+    )
+    source = models.ForeignKey(
+        'datasources.SourceSubdivision',
+        related_name='signals',
+        help_text=_('Source Subdivision'),
+        on_delete=models.PROTECT
```
Comment (krivard) [inline src/signals/models.py:227]: confirming PROTECT: the system should not let you delete a source subdivision until the signals from that subdivision have either been deleted or reassigned

i would also accept CASCADE here but i don't feel strongly either way so use your favorite

File: src/signals/models.py [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+    A model representing a signal.
+    """
+    name = models.CharField(
+        help_text=_('Name'),
+        max_length=128,
+        unique=True
+    )
+    base = models.ForeignKey(
+        'signals.Signal',
+        help_text=_('Signal base'),
```
Comment (krivard) [inline src/signals/models.py:135]: is this useful or no?

```suggestion
        related_name='base_for',
        help_text=_('Signal base'),
```

File: src/signals/models.py [hunk 16]
```diff
@@ -15,7 +15,7 @@ class TimeLabelChoices(models.TextChoices):
     A class representing choices for time labels.
     """
     DAY = 'day', _('Day')
-    DATE = 'date', _('Date')
+    DATE = 'week', _('Week')
```
Comment (krivard) [inline src/signals/models.py:18]: Unfortunately we do need all three:
* Day
* Date
* Week

**Review by krivard — CHANGES_REQUESTED** [review]
* Minor adjustment to time labels
* Exhaustively reviewed deletion methods (thanks for the reminder)
* Found a capitalization spelling i missed before (sorry)

File: src/base/models.py (hunk 9 above)
Comment (BrainIsDead) [reply src/base/models.py:13]: thanks, I just automatically copypasted

Comment (BrainIsDead) [reply src/base/models.py]: fixed

**Review by krivard — APPROVED** [review]
👍 Looks good!

**Pull request merged**